networkx/algorithms/community/demon.py

```diff
@@ -0,0 +1,296 @@
+    4:   for all C in C'(v) do:
+    5:     C <- C union v
+    6:     C' <- Merge(C', C, v)
+    7:   end for
+    8: end for
+    9: return C'
+
+    Parameters
+    ----------
+    G : graph
+        An NetworkX graph.
+
+    epsilon: real
+        The tolerance required in order to merge communities. The epsilon
+        factor is introduced to vary the percentage of common elements provided
+        from each couple of communities: epsilon = 0 ensure that two
+        communities are merged only if one of them is a proper subset of the
+        other, on the other hand with a value of epsilon = 1 even communities
+        that do not share a single node are merged together
+
+    min_community_size: integer
+        Min nodes needed to form a community
+
+    weight: String
+        If the graph is weighted, the edge parameter representing the weight
+
+    Returns
+    -------
+    communities : list
+        List of overlapping communities (sets)
+
+    References
+    ----------
+    .. [1] Michele Coscia, Giulio Rossetti, Fosca Giannotti, Dino Pedreschi:
+           DEMON: a local-first discovery method for overlapping communities.
+           KDD 2012:615-623.
+    """
+
+    for n in G.nodes():
+            G.node[n]['communities'] = [n]
+
+    all_communities = []
+
+    for ego in nx.nodes(G):
+        # EgoMinusEgo and LabelPropagation phase
+        ego_minus_ego = nx.ego_graph(G, ego, 1, False)
+        community_to_nodes = _overlapping_label_propagation(ego_minus_ego, ego, weight)
+
+        # merging phase
+        for c in community_to_nodes.keys():
+            if len(community_to_nodes[c]) > min_community_size:
+                actual_community = community_to_nodes[c]
+                all_communities = _merge_communities(all_communities, set(actual_community), epsilon)
+
+    for c in all_communities:
+        yield c
+
+
+def _overlapping_label_propagation(ego_minus_ego, ego, weight, max_iter=100):
+    """
+    Label propagation algorithm based on [2]_ used to find communities in the
+    EgoMinusEgo Graph.
+
+    Parameters
+    ----------
+    ego_minus_ego : graph
+        ego network minus its center
+
+    ego : string
+        ego network cente
+
+    weight : string
+        If the graph is weighted, the edge parameter representing the weight
+
+    max_iter : integer
+        number of desired iteration for the label propagation
+
+    Returns
+    -------
+    community_to_nodes : dict
+        List of communities in the EgoMinusEgo graph
+
+    References
+    ----------
+    .. [2] Raghavan, Usha Nandini, Réka Albert, and Soundar Kumara. "Near
+           linear time algorithm to detect community structures in large-scale
+           networks." Physical Review E 76.3 (2007): 036106.
+
+    """
+    old_node_to_coms = {}
+
+    for t in range(max_iter):
+        node_to_coms = {}
+
+        nodes = nx.nodes(ego_minus_ego)
+        random.shuffle(nodes)
+
+        for n in nodes:
+            label_freq = {}
```

Comment:
Use `defaultdict(lambda: 0)` here so that you can just do `label_freq[nn_c] += ego_minus_ego.edge[nn][n][weight] if weight else 1` below.